### PR TITLE
feat: document live migration from TimescaleDB to Timescale

### PR DESCRIPTION
**Pull request by harkishen**
# Description

This PR adds live migration documentation for TimescaleDB to Timescale migration and modifies the existing Postgres to Timescale migration documentation by using live migration docker image. The live migration image simplifies the migration process by orchestrating steps involved in CDC process, providing a seamless migration experience.

# Links

Fixes https://github.com/timescale/team-data-onboarding/issues/40

# Review checklists

Reviewers: use this section to ensure you have checked everything before approving this PR:

## Subject matter expert (SME) review checklist

*   [ ] Is the content technically accurate?
*   [ ] Is the content complete?
*   [ ] Is the content presented in a logical order?
*   [ ] Does the content use appropriate names for features and products?
*   [ ] Does the content provide relevant links to further information?

## Documentation team review checklist

*   [ ] Is the content free from typos?
*   [ ] Does the content use plain English?
*   [ ] Does the content contain clear sections for concepts, tasks, and references?
*   [ ] Have any images been uploaded to the correct location, and are resolvable?
*   [ ] If the page index was updated, are redirects required
      and have they been implemented?
*   [ ] Have you checked the built version of this content?

### CODE REVIEW

File: migrate/live-migration/index.md
Comment (alejandrodnm) [inline migrate/live-migration/index.md:41]: If they are migrating from RDS then they shouldn't have timescale installed.

**Comment by alejandrodnm** [comment]
Do we want to remove the manual instructions? I thought we were going to leave them for customers with particular use cases that are not served by the image. Maybe they need to do some extra steps, and they can leverage the manual instructions.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [inline migrate/live-migration/live-migration-from-timescaledb.md]: Wouldn't it be better to run it as a daemon?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [inline migrate/live-migration/live-migration-from-timescaledb.md]: This 2 sections talk about validation. I still think that the wait for confirmation to check the integrity is not necessary, specially since we are asking them to validate again when the migration ends.

My gripe is that we leave it for the customer to find out what validation is, which might add more stress to the process (specially since it's on the downtime phase).

What about adding telling them to validate while the data is being replicated, like run some queries with an upper bound that's shared between target and source while the migration is going. You can re-run them again at the end, using the rest of the time.

Comment (alejandrodnm) [reply migrate/live-migration/live-migration-from-timescaledb.md]: @jgpruitt what did you use to do for validation when doing replication? Did you just assume all the data was there? run some queries?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [inline migrate/live-migration/live-migration-from-timescaledb.md]: What are the `validation queries`? If we are going to mention something like this, we should explain it further, maybe some examples.

Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Validation queries is a generic term that has been described in step 4. The point here is to basically give a hint to go through the step 4 before continuing here just so that the user is prepared. The preparation is to keep the downtime duration as low as possible. We do mention about some examples in step 4, like `You could compare the number of rows or an aggregate of columns to validate that the target database matches with the source.`

**Comment by harkishen** [comment]
> Do we want to remove the manual instructions? I thought we were going to leave them for customers with particular use cases that are not served by the image. Maybe they need to do some extra steps, and they can leverage the manual instructions.

I am unclear of what advantages will it give. If users want customisation, can those customisations be done post migration? Can those customisations be done when the script pauses for `create_hypertable()`?

My understanding is that we want to give a seamless migration experience, hence keeping 1 doc for Postgres to Timescale & Timescale to Timescale will be good.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: If you mean as a background process, then this is just a recommendation. I think users may use anything they like.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: > My gripe is that we leave it for the customer to find out what validation is, which might add more stress to the process (specially since it's on the downtime phase).

I understand the point and agree to it. I want to add that we probably need a tool for verifying data across different databases, because this problem affects other migration approaches as well. Users using pg_dump & pg_restore, or dual-writes backfill will also face a same situation.

> What about adding telling them to validate while the data is being replicated, like run some queries with an upper bound that's shared between target and source while the migration is going.

This may not be right, because there are chances for DML in compressed chunks/old data that will show up as momentary inconsistencies (until it's been streamed to target).

Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: > I still think that the wait for confirmation to check the integrity is not necessary, specially since we are asking them to validate again when the migration ends.

Hmm, giving this another thought — we basically ask for 2 signals post starting live migration. They are `s` and `c` key press events.

`s` is asked because
1. We want to know at what moment is the user ready to perform the final step of migration (the step that migrates caggs from source to target). This is important as after this signal, downtime starts.
2. We give user a chance to check if initial data + live-replay data migration matches with source or not. This is mostly for safety reasons based on what I recall.

`c` is asked mainly to continue the remaining steps in migration process. These include copying sequences, caggs, and related work.

Now, the time lapse between receiving `c` and completing the migration is very small.

So, I do agree we can remove the part of validation between `s` and `c` since the time between `c` to end is relatively negligible, so if the data is not consistent, users will know it after the end of migration. Asking them to do the validation few minutes earlier (between `s` & `c`) does not make much sense.


How about this — after live migration starts, we just accept `c` only. Users will provide `c` key press when the _delay between source and target is 0_ and the _user is ready to take downtime_, meaning, we can stop live-replay and proceed with copying sequences, caggs, and remaining work.

@alejandrodnm @arajkumar WDYT?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [reply migrate/live-migration/live-migration-from-timescaledb.md]: @alejandrodnm It is a good question. Do you mean docker daemon or system service?

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
This document provides detailed instructions to migrate data from your
PostgreSQL database to a Timescale instance with minimal downtime (on the order
of a few minutes) of your production applications, using the [live migration]
strategy. To simplify the migration, we provide you with a docker image
containing all the tools and scripts that you need to perform the live
migration.
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
## 3. Run the live migration docker image
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
Next, download and run the live migration docker image:
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]: The suggested tag doesn't exist.
```suggestion
  timescale/live-migration:main
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
the target database. After migrating the schema, it will prompt you to create
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]: Stray code block
```suggestion
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]: stray
```suggestion

```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]: I find it confusing that we provide both options "pressing `c` or sending `SIGUSR1`. Why do we have both options? Can we just skip mentioning `SIGUSR1`?

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
Once the lag between the databases is below 30 megabytes, and you're ready to
take your applications offline, stop all applications which are writing to the
source database. This is the downtime phase and will last until you have
completed the validation step (4). Be sure to go through the validation step
(4) and prepare the validation queries before you enter the downtime phase to
keep the overall downtime minimal.
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
Stopping writes to the source database allows the live migration process to
finish replicating data to the target database. This will be evident when the
replication lag reduces to 0 megabytes.
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
Once the replication lag is 0, wait for a few minutes and then provide the
signal to stop the live replay by pressing key `s` or giving a `SIGUSR1`
signal, as per your log.
You should see a `Stopped` message.
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]: I don't really like the word "promote" here, it implies doing something to the database, which we're not doing.
```suggestion
## 4. Validate the data in target database and use it as new primary
```

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md]:
```suggestion
Once you are confident with the data validation, the final step is to configure
your applications to use the target database.
```

**Review by JamesGuthrie — CHANGES_REQUESTED** [review]
Initial feedback, still reviewing.

File: migrate/live-migration/index.md
Comment (JamesGuthrie) [inline migrate/live-migration/index.md]:
```suggestion
(100&nbsp;GB-10&nbsp;TB+) with low downtime (on the order of few minutes). It requires
```

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-timescaledb.md]:
```suggestion
This document provides detailed instructions to migrate data from your
TimescaleDB database (self-hosted or on [Managed Service for TimescaleDB]) to a
Timescale instance with minimal downtime (on the order of a few minutes) of
your production applications, using the [live migration] strategy. To simplify
the migration, we provide you with a docker image containing all the tools and
scripts that you need to perform the live migration.
```

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-timescaledb.md]: The suggested tag doesn't exist
```suggestion
  timescale/live-migration:main
```

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-timescaledb.md]: Do we not also need volume mapping to support resume?

**Review by JamesGuthrie — CHANGES_REQUESTED** [review]

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [inline migrate/live-migration/live-migration-from-timescaledb.md:29]: I don't think we still need to ask for 1.2 TiB as we have cleanup scripts in the background?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (JamesGuthrie) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Since we're already using docker, I would propose that we tell users to start the container with `-dit` (detached interactive tty), and that they then use `docker attach <container name>` to attach to the container. That will decouple their shell session from the migration tool.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [inline migrate/live-migration/live-migration-from-timescaledb.md:41]: Not necessarily in this PR, but we should consider adding steps on how to migrate extension from SOURCE to TARGET.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [inline migrate/live-migration/live-migration-from-timescaledb.md]: Should the type be "warning"?

File: migrate/live-migration/live-migration-from-postgres.md
Comment (arajkumar) [reply migrate/live-migration/live-migration-from-postgres.md]: +1, since we advocate to use docker in interactive mode(-it), there is no need to mention about signals.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [reply migrate/live-migration/live-migration-from-timescaledb.md]: +1. Additionally I recommend to pass `--name live-migration`.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: @arajkumar and I discussed about this, and we agreed that we should allow users to check data consistency at the end of migration process. ATM, user have to validate data twice — first, when he gives `s`, he checks for existing data + live replay, but without caggs (since they are yet to be migrated); second time, after migration completes, he verifies for the entire database. With the new approach, user will only perform verification on entire database after the migration completes. Hence, we will keep `c`, and remove the `s` signal.

For instance, the new workflow will be as follows.

User starts live migration for TS to TS. When the delay between source and target is 0 (he may have to stop load on the source side) and the user is ready to take the "minimal downtime", he will press `c` key. The script will then stop live-replay and do the remaining work to complete the migration process. 

If the migration is for PG to TS, an additional step for `create_hypertable()` will be asked before the `c` key press.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Yeah, I meant detached `-d`

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [inline migrate/live-migration/live-migration-from-timescaledb.md]: Again, I think the phrase

> prepare the validation queries before you enter the downtime phase

Is confusing. If I had zero context I would imagine there are some validation scripts somewhere in the instructions.

If we want the customer to have something prepare for validation, we need to specify at the beginning of the instructions, not at the end. It could be an optional step of 'Prepare validation scripts'.

**Comment by alejandrodnm** [comment]
> > Do we want to remove the manual instructions? I thought we were going to leave them for customers with particular use cases that are not served by the image. Maybe they need to do some extra steps, and they can leverage the manual instructions.
> 
> I am unclear of what advantages will it give. If users want customisation, can those customisations be done post migration? Can those customisations be done when the script pauses for `create_hypertable()`?
> 
> My understanding is that we want to give a seamless migration experience, hence keeping 1 doc for Postgres to Timescale & Timescale to Timescale will be good.

That's the thing, we don't know what crazy things they might ask, they might want to exclude some tables or whatever. We can always say, if the tool doesn't fit the use case you still have the manual instructions.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Update: With https://github.com/timescale/pgcopydb/pull/22 merged, the live migration image behaviour aligns with the above mentioned approach.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md:29]: Yes, we have cleanup scripts in background, but I don't have numbers with cleanup script. However, if the data copy takes 2 days to complete, maybe its worth having 1.2TBs?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Since we don't have validation scripts ATM, and they are creating confusion, how about we remove the part ` and prepare the validation queries ...`?

Comment (alejandrodnm) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Let's do that.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: Based on my understanding, I don't think we should suggest detached mode. If we suggest `-dit`, then we will have to add about attaching the container and reading when to send signal based on the delay. Suggesting simply `screen/tmux` covers it up.

Thoughts?

File: migrate/live-migration/live-migration-from-postgres.md
Comment (JamesGuthrie) [inline migrate/live-migration/live-migration-from-postgres.md:73]: This instruction is too late. The previous steps refer to `$SOURCE`. Also, I would prefer it if we could provide an actual url as an example, instead of an empty string. Compare with the instructions in the note for dual-write and backfill: https://docs.timescale.com/migrate/latest/dual-write-and-backfill/dual-write-from-timescaledb/#3-set-up-schema-and-migrate-relational-data-to-target-database

image

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (JamesGuthrie) [reply migrate/live-migration/live-migration-from-timescaledb.md]: @Harkishen-Singh This change hasn't been made yet, or am I missing something?

Comment (JamesGuthrie) [reply migrate/live-migration/live-migration-from-timescaledb.md]: > have to add about attaching the container and reading when to send signal based on the delay
> ...
> Suggesting simply screen/tmux covers it up.

But suggesting it doesn't mean that people will do it, or that they know how to do it. Why would we need to give the user instructions on how to do it with Docker, but we don't have to give them instructions if they're using screen/tmux?

File: _partials/_migrate_live_migration_step2.md
Comment (JamesGuthrie) [inline _partials/_migrate_live_migration_step2.md]: This duplicated the content which is already in a partial, `./_partials/_migrate_set_up_source_and_target.md` please use that instead.

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (alejandrodnm) [reply migrate/live-migration/live-migration-from-timescaledb.md]: BTW I've just realize you can use `docker run` and detach with CTRL-P+Ctrl-Q

https://docs.docker.com/engine/reference/commandline/cli/#default-key-sequence-to-detach-from-containers

**Review by JamesGuthrie — APPROVED** [review]
I won't be able to give a proper final review, but I don't want to block the merge, so I'm approving.

**Comment by harkishen** [comment]
I will wait for other approvals before merging. We should also probably wait for product for the target launch date before merging.

**Comment by JamesGuthrie** [comment]
I think that we should merge docs as soon as they're ready. @VineethReddy02 WDYT?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [reply migrate/live-migration/live-migration-from-timescaledb.md:29]: Ahh, I forgot about that :D

**Review by alejandrodnm — APPROVED** [review]

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (arajkumar) [inline migrate/live-migration/live-migration-from-timescaledb.md]: It is better to be explit about the host directory which is being mounted to docker must have enough disk space.

```suggestion
  -v ~/live-migration:/opt/timescale/ts_cdc \
```

**Comment by VineethReddy02** [comment]
@Harkishen-Singh is this PR ready to be merged?

File: migrate/live-migration/live-migration-from-timescaledb.md
Comment (harkishen) [reply migrate/live-migration/live-migration-from-timescaledb.md]: I have also added a note below this line

```
Ensure `/live-migration` directory has sufficient space available.
Ideally, 1.5 times the source database size should be good.
```

**Comment by harkishen** [comment]
@VineethReddy02 yes.

**Review by arajkumar — APPROVED** [review]

**Pull request merged**
